### PR TITLE
Unpacking Containers

### PR DESCRIPTION
**Pull request by bobturneruk**
I'm planning to write a post on virtual machines and containers, somewhat to force me to get my head round them better.

Any and all feedback appreciated. I could easily have got some of this factually wrong as I'm still quite new to it!

### CODE REVIEW

File: _posts/2019-10-15-unpacking-containers.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+description:
+type: text
+excerpt_separator: <!--more-->
+---
+
+I worked in commercial I.T. for a few years before doing an PhD and spending many years in scientific research. Coming back to software full time is a bit like coming out of cryo-stasis: Technology has moved on in ways I could not possibly have imagined and I must adapt like some kind of sci-fi person. One of the big changes over the last few years has been the increased use of virtual machines (e.g. using [Virtualbox](https://www.virtualbox.org)) and containers ([Docker](https://www.docker.com/) is a popular choice). I now feel ready to talk about this with others.
+
+<!--more-->
+
+If you've persevered to this point in the point, you deserve an explaination of what virtual machines and containers actually are. But you are going to have to wait until I've explained how software interacts with "normal" hardware such as a laptop.
```
Comment (drj11) [inline _posts/2019-10-15-unpacking-containers.md]: "this point in the point" doesn't make sense. do you mean "this point in the post" or something?

Comment (davidwyld) [inline _posts/2019-10-15-unpacking-containers.md]: post rather than point?
explanantion*

File: _posts/2019-10-15-unpacking-containers.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+{% include image_caption.html url="/assets/images/vm-container-blog/no_vm.png" description="Computers, operating systems and software"%}
+
+When software is installed, typically a bunch of files are copied onto the computer and some other files are altered. Yet more prerequisite files might be downloaded from another source. A problem arises if, by installing one piece of software, a file on which another piece of software depends is altered or removed. This can be accidental or malicious. Malicious software can be deliberately designed to access data which should remain private to other software.
+
+Another potential problem is a situation where the software runs, but behaves differently on different computers. This could involve the software crashing, or perhaps producing slightly different results - a big problem when trying to make results reproducible.
```
Comment (drj11) [inline _posts/2019-10-15-unpacking-containers.md]: replace " - " with "—" (that's unicode m-dash, U+2014)

Comment (drj11) [reply _posts/2019-10-15-unpacking-containers.md]: there may even be some clever way to get m-dashes in markdown without having the memorise the hex-codes for unicode characters.

File: _posts/2019-10-15-unpacking-containers.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Virtual machines...
+
+A virtual machine (or VM) is, in effect, a piece of software running on a physical computer, that acts like another operating system. A computer in the imagination of another.
```
Comment (drj11) [inline _posts/2019-10-15-unpacking-containers.md]: "imagination" is a nice metaphor.

File: _posts/2019-10-15-unpacking-containers.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+- **Software that runs on one of these computers will run on any another, producing the same results (and crashing in the same way).**
+- Software on one virtual machine can't easily interact with (mess up) another.
+- If we need a better virtual machine, we can tell it to take up more of the resources of the physical computer on which its running.
+- If that doesn't work, we can copy the virtual machine onto (or recreate it exactly on) a more powerful physical computer, which could be on the other side of the world.
+
+So virtual machines can be really helpful, but they can be slow to run and also slow to configure.
+
+## Containers...
+
+Containers get round some of the problems with virtual machines. The term "container" is not always defined in exactly the same way, but in general: Rather than running on an operating system like virtual machines, containers allow software running in them access hardware more directly, but restrict what hardware they can use and interaction between discrete containers. This means they avoid some of the slowdown associated with virtual machines, but can still be configured and replicated to ensure that software runs consistently.
```
Comment (drj11) [inline _posts/2019-10-15-unpacking-containers.md]: this is either inaccurate or a poor way of describing it. But I sympathise, the whole area is littered with poor descriptions and poorly defined terms. Software that runs on a VM, conventionally, has to rely on the VM having its own operating system (which they do). Containers do not have their own operating system, they rely on the operating system of the host.

File: _posts/2019-10-15-unpacking-containers.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+So far, so good. But probably not good enough to justify the current buzz around these technologies. I think one of the key benefits that we've not mentioned is how virtual machines and containers can be set up to work together.
+
+[Vagrant](https://www.vagrantup.com/), for example, can be used to create and provision (install software, write configuration files, mount network drives, etc.) any number of virtual machines with the following:
+
+```sh
+vagrant up
+```
+
+Easy, eh? Every aspect of the configuration is defined as software (this is "*infrastructure as software*"). I use this to configure linked database server - web server pairs. I can use the same approach to configure virtual machines on my laptop for testing, for development servers and for live applications. Doing this with physical hardware would be epic, and the configurations would drift apart through updates and bug fixes in milliseconds (not really, but quickly though).
```
Comment (drj11) [inline _posts/2019-10-15-unpacking-containers.md]: again with the m-dash

**Review by drj11 — COMMENTED** [review]
If you have the energy, can you link to Singularity too? https://docs.hpc.shef.ac.uk/en/latest/sharc/software/apps/singularity.html

Otherwise, Looks Good To Me!

**Review by drj11 — APPROVED** [review]

**Comment by drj11** [comment]
Hmm. Possibly a comment I made earlier has been lost because I can't drive the GitHub web app properly.

You're good to merge as-is, but you could also consider dropping in a link to Singularity on ShARC: https://docs.hpc.shef.ac.uk/en/latest/sharc/software/apps/singularity.html

File: _posts/2019-10-15-unpacking-containers.md (hunk 1 above)
Comment (bobturneruk) [reply _posts/2019-10-15-unpacking-containers.md]: Fixed.

File: _posts/2019-10-15-unpacking-containers.md (hunk 3 above)
Comment (bobturneruk) [reply _posts/2019-10-15-unpacking-containers.md]: Thanks!

File: _posts/2019-10-15-unpacking-containers.md (hunk 2 above)
Comment (bobturneruk) [reply _posts/2019-10-15-unpacking-containers.md]: Done.

File: _posts/2019-10-15-unpacking-containers.md (hunk 5 above)
Comment (bobturneruk) [reply _posts/2019-10-15-unpacking-containers.md]: Changed.

**Pull request merged**